cmd/metal-api/internal/auditing/auditing-interceptor.go

```diff
@@ -0,0 +1,162 @@
+
+func UnaryServerInterceptor(a Auditing, logger *zap.SugaredLogger, shouldAudit func(fullMethod string) bool) grpc.UnaryServerInterceptor {
+	if a == nil {
+		return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+			return handler(ctx, req)
+		}
+	}
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		if !shouldAudit(info.FullMethod) {
+			return handler(ctx, req)
+		}
+		correlationID := uuid.New().String()
+		childCtx := context.WithValue(ctx, auditingCorrelationIDKey, correlationID)
+		err = a.Index("correlation-id", correlationID, "method", info.FullMethod, "kind", "grpc", "req", req)
+		if err != nil {
+			return nil, err
+		}
+		resp, err = handler(childCtx, req)
+		if err != nil {
+			err2 := a.Index("correlation-id", correlationID, "method", info.FullMethod, "kind", "grpc", "err", err)
+			if err2 != nil {
+				logger.Errorf("unable to index error: %v", err2)
+			}
+			return nil, err
+		}
+		err = a.Index("correlation-id", correlationID, "method", info.FullMethod, "kind", "grpc", "resp", resp)
+		return resp, err
+	}
+}
+
+func StreamServerInterceptor(a Auditing, logger *zap.SugaredLogger, shouldAudit func(fullMethod string) bool) grpc.StreamServerInterceptor {
+	if a == nil {
+		return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+			return handler(srv, ss)
+		}
+	}
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		if !shouldAudit(info.FullMethod) {
+			return handler(srv, ss)
+		}
+		correlationID := uuid.New().String()
+		err := a.Index("kind", "grpc-stream", "stream", srv)
+		if err != nil {
+			return err
+		}
+		err = handler(srv, ss)
+		if err != nil {
+			err2 := a.Index("correlation-id", correlationID, "method", info.FullMethod, "kind", "grpc-stream", "err", err)
+			if err2 != nil {
+				logger.Errorf("unable to index error: %v", err2)
+			}
+			return err
+		}
+		err = a.Index("correlation-id", correlationID, "method", info.FullMethod, "kind", "grpc-stream")
+		return err
+	}
+}
+
+func HttpMiddleware(a Auditing, logger *zap.SugaredLogger, shouldAudit func(*url.URL) bool) func(h http.Handler) http.Handler {
+	if a == nil {
+		return func(h http.Handler) http.Handler {
+			return h
+		}
+	}
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			switch r.Method {
+			case http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete:
+				break
+			default:
+				h.ServeHTTP(w, r)
+				return
+			}
+			if !shouldAudit(r.URL) {
+				h.ServeHTTP(w, r)
+				return
+			}
+			correlationID := uuid.NewString()
+			r = r.WithContext(context.WithValue(r.Context(), auditingCorrelationIDKey, correlationID))
```

Comment:
The logger filter chain also injects a context into the request. Maybe it makes sense to reuse it in order to align logs with audit traces: https://github.com/metal-stack/metal-lib/blob/master/rest/middleware.go#L79-L80

Maybe it's also beneficial to return a `restful.FilterFunction` directly from this function.